cyral/resource_cyral_integration_idp_saml_draft.go

```diff
@@ -0,0 +1,281 @@
+//
+// 1. Provide the SP metadata generated by this resource to the IdP.
+//
+// 2. Provide the IdP metadata to the `cyral_integration_idp_saml` resource.
+//
+
+type CreateGenericSAMLDraftRequest struct {
+	DisplayName              string                  `json:"displayName"`
+	DisableIdPInitiatedLogin bool                    `json:"disableIdPInitiatedLogin"`
+	IdpType                  string                  `json:"idpType,omitempty"`
+	Attributes               *RequiredUserAttributes `json:"attributes,omitempty"`
+}
+
+func (req *CreateGenericSAMLDraftRequest) ReadFromSchema(d *schema.ResourceData) error {
+	req.DisplayName = d.Get("display_name").(string)
+	req.DisableIdPInitiatedLogin = d.Get("disable_idp_initiated_login").(bool)
+	req.IdpType = d.Get("idp_type").(string)
+
+	attributes, err := RequiredUserAttributesFromSchema(d)
+	if err != nil {
+		return err
+	}
+	req.Attributes = attributes
+
+	return nil
+}
+
+type GenericSAMLDraftResponse struct {
+	Draft GenericSAMLDraft `json:"draft"`
+}
+
+func (resp *GenericSAMLDraftResponse) WriteToSchema(d *schema.ResourceData) error {
+	d.SetId(resp.Draft.ID)
+	if err := d.Set("display_name", resp.Draft.DisplayName); err != nil {
+		return err
+	}
+	if err := d.Set("disable_idp_initiated_login", resp.Draft.DisableIdPInitiatedLogin); err != nil {
+		return err
+	}
+	if err := d.Set("sp_metadata", resp.Draft.SPMetadata.XMLDocument); err != nil {
+		return err
+	}
+	if err := d.Set("idp_type", resp.Draft.IdpType); err != nil {
+		return err
+	}
+	if resp.Draft.Attributes != nil && len(d.Get("attributes").(*schema.Set).List()) > 0 {
+		if err := resp.Draft.Attributes.WriteToSchema(d); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type ListGenericSAMLDraftsResponse struct {
+	Drafts []GenericSAMLDraft `json:"drafts"`
+}
+
+func CreateGenericSAMLDraftConfig() ResourceOperationConfig {
+	return ResourceOperationConfig{
+		Name:       "GenericSAMLDraftResourceCreate",
+		HttpMethod: http.MethodPost,
+		CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+			return fmt.Sprintf("https://%s/v1/integrations/generic-saml/drafts", c.ControlPlane)
+		},
+		NewResourceData: func() ResourceData { return &CreateGenericSAMLDraftRequest{} },
+		NewResponseData: func() ResponseData { return &GenericSAMLDraftResponse{} },
+	}
+}
+
+func ReadGenericSAMLDraftConfig() ResourceOperationConfig {
+	return ResourceOperationConfig{
+		Name:       "GenericSAMLDraftResourceRead",
+		HttpMethod: http.MethodGet,
+		CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+			return fmt.Sprintf("https://%s/v1/integrations/generic-saml/drafts/%s", c.ControlPlane, d.Id())
+		},
+		HandleRequestError: func(d *schema.ResourceData, c *client.Client, err error) error {
+			httpError, ok := err.(*client.HttpError)
+			if !ok || httpError.StatusCode != http.StatusNotFound {
+				return err
+			}
+			log.Printf("[DEBUG] SAML draft not found. Checking if completed draft exists.")
```

Comment:
I did try to set `id` in the past, and it didn't work. The reason it works for the S3 bucket case is because `id` is a user-settable argument (i.e. `Optional: true`). I am not sure if that makes sense in our case, since the user should have no control on the ID. What do you think?